### ISSUE TITLE
Data wipe and duplication while simultaneously indexing items in auto-batch

### ISSUE THREAD

**Issue by DavertMik** [meilisearch/meilisearch#2470]
### TLDR

*I found the root cause of issue!*

My guess is: Meilisearch fails to index items when individual items are added simultaneously in auto-batch mode on 1 core CPU.

I'm running 1 CPU AMD (can this be the cause?) and when I send a few documents at once to Meilisearch it creates items as duplicates.

Every time I do any indexing in my app I'm synchronizing indexing using `wait_for_task` from Ruby client, to ensure no new batch of documents are sent after previous completed processing. If I run a few indexing tasks in parallel I get data loss or a broken index (see below).

The following scenario describes how the data loss happens and the point where internal error  `"message":"internal: Missing key in the documents database..` can be seen

### Scenario

I launched Meilisearch in auto-batch mode:

```
/usr/bin/meilisearch --db-path /var/lib/meilisearch/data1.ms --env production --dumps-dir /var/lib/meilisearch/dumps --enable-auto-batching
```

Here is my index:

```
{
  "uid": "Test_staging",
  "name": "Test_staging",
  "createdAt": "2022-06-06T12:44:50.221884241Z",
  "updatedAt": "2022-06-06T17:41:11.056253545Z",
  "primaryKey": "id"
}
```
here are settings for it:

```
{
  "displayedAttributes": [
    "*"
  ],
  "searchableAttributes": [
    "title_tags",
    "public_title",
    "file",
    "description"
  ],
  "filterableAttributes": [
    "assigned_to",
    "branch_id_number",
    "created_at",
    "priority",
    "project_id",
    "state",
    "sync",
    "updated_at"
  ],
  "sortableAttributes": [
    "branch_id_number"
  ],
  "rankingRules": [
    "branch_id_number:desc",
    "words",
    "typo",
    "attribute",
    "exactness"
  ],
  "stopWords": [],
  "synonyms": {},
  "distinctAttribute": "public_uid",
  "typoTolerance": {
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9
    },
    "disableOnWords": [],
    "disableOnAttributes": []
  }
}
```

Then I added 2 documents **simultaneously but not in batch** (I assume this causes an issue).

These documents:
```json
[
{"id": 228142, "title": "asdsad", "state": "automated", "priority": "normal", "public_uid": "37ccf021", "project_id": 78207, "branch_id_number": 0}, 
{"id": 228143, "title": "something", "state": "automated", "priority": "normal", "public_uid": "39c6499b", "project_id": 78207, "branch_id_number": 0 }
]
```
One with the title "asdsad" other with title 'something' on different IDs

But Meilisearch contained the same document under both IDs. So I tried to re-add them via CURL:

I add 2 documents via CURL:

```
curl -X POST 'http://localhost:7700/indexes/Test_staging/documents?primaryKey=id' -H 'Content-Type: application/json' --data-binary '[{"id": 228142, "title": "asdsad", "state": "automated", "priority": "normal", "public_uid": "37ccf021", "project_id": 78207, "branch_id_number": 0}, {"id": 228143, "title": "something", "state": "automated", "priority": "normal", "public_uid": "39c6499b", "project_id": 78207, "branch_id_number": 0 }]' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   432  100   130  100   302  21666  50333 --:--:-- --:--:-- --:--:-- 72000
{
  "uid": 3894,
  "indexUid": "Test_staging",
  "status": "enqueued",
  "type": "documentAddition",
  "enqueuedAt": "2022-06-06T17:41:04.796354715Z"
}
```

They are added

```
curl -X GET 'http://localhost:7700/tasks/3894' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   321  100   321    0     0  53500      0 --:--:-- --:--:-- --:--:-- 64200
{
  "uid": 3894,
  "indexUid": "Test_staging",
  "status": "succeeded",
  "type": "documentAddition",
  "details": {
    "receivedDocuments": 2,
    "indexedDocuments": 4
  },
  "duration": "PT3.013141648S",
  "enqueuedAt": "2022-06-06T17:41:04.796354715Z",
  "startedAt": "2022-06-06T17:41:10.341110888Z",
  "finishedAt": "2022-06-06T17:41:13.354252536Z",
  "batchUid": 3893
}
```
But in fact **the first document only** is stored, and the second one is ignored.
There are documents named "something" and "adsad" I query both of them but in both cases, I receive the document #1

```
curl  -X GET 'http://localhost:7700/indexes/Test_staging/documents/228142' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   137  100   137    0     0  22833      0 --:--:-- --:--:-- --:--:-- 22833
{
  "id": 228143,
  "title": "something",
  "state": "automated",
  "priority": "normal",
  "public_uid": "39c6499b",
  "project_id": 78207,
  "branch_id_number": 0
}
curl -X GET 'http://localhost:7700/indexes/Test_staging/documents/228143' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   137  100   137    0     0  27400      0 --:--:-- --:--:-- --:--:-- 27400
{
  "id": 228143,
  "title": "something",
  "state": "automated",
  "priority": "normal",
  "public_uid": "39c6499b",
  "project_id": 78207,
  "branch_id_number": 0
}
``` 

I tried to write documents to new IDs, when I use a new ID which (in my case were 728142 and 728143) and in this case documents **were store correctly**. These IDs were not exist in database.

So my guess was that overwriting current data can be a problem.

Then I tried to delete items 228142 and 228143 from index and then re-add them:

```json
{"uid":4011,"indexUid":"Test_staging","status":"succeeded","type":"documentDeletion","details":{"receivedDocumentIds":1,"deletedDocuments":1},"duration":"PT0.775224943S","enqueuedAt":"2022-06-06T20:16:51.462232736Z","startedAt":"2022-06-06T20:16:51.464172707Z","finishedAt":"2022-06-06T20:16:52.23939765Z","batchUid":4011}root@meilisearch0271ondebian10buster-s-1vcpu-2gb-amd-ams3-01:~# 
```

But when I try to add new item I receive this **internal** error:

```
{"uid":4012,"indexUid":"Test_staging","status":"failed","type":"documentAddition","details":{"receivedDocuments":2,"indexedDocuments":0},"error":{"message":"internal: Missing key in the documents database..","code":"internal","type":"internal","link":"https://docs.meilisearch.com/errors#internal"},"duration":"PT0.003221089S","enqueuedAt":"2022-06-06T20:16:58.45716101Z","startedAt":"2022-06-06T20:16:58.
```

```js
-X GET 'http://localhost:7700/indexes/Test_staging/documents/928142' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    94  100    94    0     0  18800      0 --:--:-- --:--:-- --:--:-- 18800
{
  "id": 928142,
  "title": "asdsad",
  "state": "automated",
  "priority": "normal",
  "public_uid": "37ccf021"
}
 curl -X GET 'http://localhost:7700/indexes/Test_staging/documents/928143' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    97  100    97    0     0  19400      0 --:--:-- --:--:-- --:--:-- 19400
{
  "id": 928143,
  "title": "something",
  "state": "automated",
  "priority": "normal",
  "public_uid": "39c6499b"
}

``` 

**Meilisearch version:** 0.27.1

Stats for this index:

```
{
  "numberOfDocuments": 99101,
  "isIndexing": false,
  "fieldDistribution": {
    "assigned_to": 99100,
    "branch_id": 99100,
    "branch_id_number": 99101,
    "changed_in_branch": 99100,
    "cid": 99100,
    "code": 99100,
    "created_at": 99100,
    "created_by": 99100,
    "created_in_branch": 99100,
    "deleted_in_branch": 99100,
    "description": 99100,
    "emoji": 99100,
    "environment": 99100,
    "file": 99100,
    "has_examples": 99100,
    "id": 99101,
    "import_id": 99100,
    "jira_link": 99100,
    "last_sync_id": 99100,
    "params": 99100,
    "position": 99100,
    "precondition": 99100,
    "priority": 99101,
    "project_id": 99101,
    "public_title": 99100,
    "public_uid": 99101,
    "recordings_count": 99100,
    "run_statuses": 99100,
    "state": 99101,
    "suite_id": 99100,
    "sync": 99100,
    "tags": 99100,
    "title": 99101,
    "title_tags": 99100,
    "updated_at": 99100
  }
}
```

* To create index I used Ruby on Rails integration and Ruby client.
* All interactions were re-played over CURL to demonstrate the issue.
* I tried to change the primary key index from ID to a compound index "#{project_id}_#{public_uid}" and recreate index. I get absolutely the same behavior.
* can this be that auto-batching breaks the atomic additions of documents, combining them in a batch even they were executed as separate requests?
* Meilisearch is running on DigitalOcean droplet created via Marketplace

**Comment by Kerollmops** [comment]
Hey @DavertMik,

Thank you very much for the detailed reproduction steps, I am happy to be able to reproduce it on my side and I am currently trying to find and fix the issue. The bug is also on the _main_ branch not only _v0.27.1_.

Here is a simpler version of the steps to reproduce the bug, the first block of code is the list of documents you gave us and we store them in a _documents.json_ file.

```json
[
    {
        "id": 228142,
        "title": "asdsad",
        "state": "automated",
        "priority": "normal",
        "public_uid": "37ccf021",
        "project_id": 78207,
        "branch_id_number": 0
    },
    {
        "id": 228143,
        "title": "something",
        "state": "automated",
        "priority": "normal",
        "public_uid": "39c6499b",
        "project_id": 78207,
        "branch_id_number": 0
    }
]
```

We run the server with a debounce duration just to make sure that the auto-batching system creates a new batch from your two documents that were sent independently. If we don't add a debounce the auto-batching system just processes the two updates independently and everything works fine.

```bash
cargo run --release -- --no-analytics --enable-auto-batching --debounce-duration-sec 5 --log-level=debug
```

Now that the engine run just sends the two documents independently one just after the other, you have 5sec to send the second document.

```bash
cat documents.json | jq '.[0]' | xh 'http://localhost:7700/indexes/repro/documents?primaryKey=id' content-type:application/json
cat documents.json | jq '.[1]' | xh 'http://localhost:7700/indexes/repro/documents?primaryKey=id' content-type:application/json
```

Once the update has been processed, the first document that has been sent is displayed as the second. The bug is surely located in either the auto-batching system of Meiliseqrch or in the update transform system of milli.

**Comment by DavertMik** [comment]
@Kerollmops thank you very much for the quick look and for reproducing the issue. I spent 4 days catching it and that's so nice my efforts were meaningful

>  --debounce-duration-sec 5 

So what is a quick fix for this error? Can I disable debounce like this so two updates were running independently?

```
--debounce-duration-sec 0
```

**Comment by DavertMik** [comment]
@irevoire I tried, 4 secs for indexing 2 documents is too slow for me.
Everything gets really really slow without auto-batching ... 🤢

so if no quick fix I can just update my application to combine all batched updates to be sent as 1 request

**Comment by irevoire** [comment]
Hey @DavertMik,

Instead of releasing a quick fix, we released the real fix quickly :racing_car: 
The v0.27.2 is on its way; it's already available in Brew. See the following [gh-action](https://github.com/meilisearch/meilisearch/actions/runs/2455319183) to follow the release of the Debian package.

If the issue persists, please re-open the issue.

Thank you very much for the report, have a nice day.

**Comment by DavertMik** [comment]
Thank you so much for the FASTEST fix!

**Comment by DavertMik** [comment]
Is there an estimate for the Docker release?

**Comment by curquiza** [comment]
Hello @DavertMik, sorry we are having unexpected issues with our docker CIs to build and push the docker images.
We are working on it, sorry for the inconveniance
